c/storageApiml.c

```diff
@@ -0,0 +1,648 @@
+  }
+  if (response->_httpClientContext) {
+    httpClientContextDestroy(response->_httpClientContext);
+  }
+  safeFree((char*)response, sizeof(*response));
+}
+
+static int transformHttpStatus(int httpStatus) {
+  switch (httpStatus) {
+    case HTTP_STATUS_NOT_FOUND:
+      return STORAGE_STATUS_KEY_NOT_FOUND;
+    case HTTP_STATUS_FORBIDDEN:
+      return STORAGE_STATUS_INVALID_CLIENT_CERT;
+    case HTTP_STATUS_OK:
+      return STORAGE_STATUS_OK;
+    case HTTP_STATUS_NO_CONTENT:
+      return STORAGE_STATUS_OK;
+    case HTTP_STATUS_CREATED:
+      return STORAGE_STATUS_OK;
+    default:
+      return STORAGE_STATUS_HTTP_ERROR;
+  }
+}
+
+static ApimlResponse *apimlDoRequest(ApimlStorage *storage, ApimlRequest *request, int *statusOut) {
+  int status = 0;
+  int statusCode = 0;
+  TlsEnvironment *tlsEnv = storage->tlsEnv;
+  HttpClientSettings *clientSettings = storage->clientSettings;
+  HttpClientContext *httpClientContext = NULL;
+  HttpClientSession *session = NULL;
+  LoggingContext *loggingContext = makeLoggingContext();
+  char *path = request->path;
+  char *body = request->body;
+  int bodyLen = request->bodyLen;
+  Json *jsonResponse = NULL;
+  char *pluginId = (char*)storage->pluginId;
+  zowelog(NULL, LOG_COMP_ID_APIML_STORAGE, ZOWE_LOG_DEBUG, "apiml request %s %s\n", request->method, request->path);
+
+  do {
+    status = httpClientContextInitSecure(clientSettings, loggingContext, tlsEnv, &httpClientContext);
+    if (status) {
+      zowelog(NULL, LOG_COMP_ID_APIML_STORAGE, ZOWE_LOG_DEBUG, "error in httpcb ctx init: %d\n", status);
+      *statusOut = STORAGE_STATUS_HTTP_ERROR;
+      break;
+    }
+    status = httpClientSessionInit(httpClientContext, &session);
+    if (status) {
+      zowelog(NULL, LOG_COMP_ID_APIML_STORAGE, ZOWE_LOG_DEBUG, "error initing session: %d\n", status);
+      *statusOut = STORAGE_STATUS_HTTP_ERROR;
+      break;
+    }
+    status = httpClientSessionStageRequest(httpClientContext, session, request->method, path, NULL, NULL, body, bodyLen);
+    if (status) {
+      zowelog(NULL, LOG_COMP_ID_APIML_STORAGE, ZOWE_LOG_DEBUG, "error staging request: %d\n", status);
+      *statusOut = STORAGE_STATUS_HTTP_ERROR;
+      break;
+    }
+    requestStringHeader(session->request, TRUE, "Content-type", "application/json");
+    requestStringHeader(session->request, TRUE, "X-CS-Service-ID", pluginId);
+    status = httpClientSessionSend(httpClientContext, session);
+    if (status) {
+      zowelog(NULL, LOG_COMP_ID_APIML_STORAGE, ZOWE_LOG_DEBUG, "error sending request: %d\n", status);
+      *statusOut = STORAGE_STATUS_HTTP_ERROR;
+      break;
+    }
+    if (request->parseJson) {
+      jsonResponse = receiveJsonResponse(httpClientContext, session, &status);
+    } else {
+      receiveResponse(httpClientContext, session, &status);
+    }
+    if (status) {
+      *statusOut = status;
+      break;
+    }
+  } while (0);
+  if (status) {
+    if (session) {
+      httpClientSessionDestroy(session);
```

Comment:
Looks like this is only being done in the case of failure. Is this correct or a leak?

Reply:
Nevermind, its handled in freeApimlResponse